src/azure-cli/azure/cli/command_modules/role/custom.py

```diff
@@ -854,37 +836,77 @@ def list_permission_grants(cmd, identifier=None, query_filter=None, show_resourc
-    if existing_resource_access:
-        existing_resource_access.resource_access += resource_accesses
-    else:
-        required_resource_access = RequiredResourceAccess(resource_app_id=api,
-                                                          resource_access=resource_accesses)
-        existing.append(required_resource_access)
-    update_parameter = ApplicationUpdateParameters(required_resource_access=existing)
+    update_parameter = ApplicationUpdateParameters(required_resource_access=rras)
     graph_client.applications.patch(application.object_id, update_parameter)
-    logger.warning('Invoking "az ad app permission grant --id %s --api %s" is needed to make the '
-                   'change effective', identifier, api)
 
 
-def delete_permission(cmd, identifier, api):
+def remove_required_permission(cmd, identifier, api, api_permissions=None):
+    """Removes a required permissions for an API from requiredResourceAccess
+
+    :param api_permissions: If given, only these required permissions will be removed. If
+            omitted, all required permissions for the API will be removed.
+    """
+
     graph_client = _graph_client_factory(cmd.cli_ctx)
+    api_app_id = _resolve_api_identifier_to_app_id(graph_client, api)
     application = show_application(graph_client.applications, identifier)
-    existing_accesses = application.required_resource_access
-    existing_accesses = [e for e in existing_accesses if e.resource_app_id != api]
-    update_parameter = ApplicationUpdateParameters(required_resource_access=existing_accesses)
+    rras = application.required_resource_access
+
+    updated_rras = []
+    for rra in rras:
+        if rra.resource_app_id == api_app_id:
+            if api_permissions is not None:
+                # If specific permissions were given, include the RRA only if there are any remaining
+                # permissions after filtering out the permissions to remove.
+                for api_permission in api_permissions:
+                    permission_id, permission_type = api_permission.split('=')
+                    rra.resource_access = [ra for ra in rra.resource_access
+                                           if ra.type != permission_type or ra.id != permission_id]
+                if rra.resource_access:
+                    updated_rras.append(rra)
+        else:
+            updated_rras.append(rra)
+
+    update_parameter = ApplicationUpdateParameters(required_resource_access=updated_rras)
     return graph_client.applications.patch(application.object_id, update_parameter)
 
 
+def _resolve_api_identifier_to_app_id(graph_client, api_identifier):
+    """Resolves an API identifier to an app ID
+
+    The identifier can be an object ID or service principal name for an existing service principal
+    object. If no matching service principal object is found but the identifier is a Guid, it is
+    assumed to be the app ID of an API which is not instantiated in the tenant.
```

Comment:
```suggestion
    The identifier can be an object ID or service principal name, like `https://graph.microsoft.com`
    for an existing service principal object.
    If no matching service principal object is found but the identifier is a Guid, it is
    assumed to be the app ID of an API which is not instantiated in the tenant.
```

Reply:
As a side note, this is the value of `service_principal_names` for `api_identifier='https://graph.microsoft.com'`:

```
<class 'list'>: ['https://dod-graph.microsoft.us', 'Microsoft.Azure.AgregatorService', 
'00000003-0000-0000-c000-000000000000', '00000003-0000-0000-c000-000000000000/ags.windows.net',
'https://ags.windows.net', 'https://graph.microsoft.com', 'https://canary.graph.microsoft.com', 
'https://graph.microsoft.com/', 'https://graph.microsoft.us']
```

Reply:
Added example value.